reV/bespoke/gradient_free.py

```diff
@@ -0,0 +1,247 @@
+        for i in range(self.population_size):
+            self.chromosome_2_variables(self.parent_population[i])
+            self.parent_fitness[i] = \
+                self.objective_function(self.design_variables)
+
+    def chromosome_2_variables(self, chromosome):
+        """convert the binary chromosomes to design variable values"""
+
+        first_bit = 0
+        float_ind = 0
+
+        for i in range(self.nvars):
+            binary_value = 0
+            for j in range(self.bits[i]):
+                binary_value += chromosome[first_bit + j] * 2**j
+            first_bit += self.bits[i]
+
+            if self.variable_type[i] == "float":
+                self.design_variables[i] = \
+                    self.discretized_variables["float_var%s"
+                                               % float_ind][binary_value]
+                float_ind += 1
+
+            elif self.variable_type[i] == "int":
+                self.design_variables[i] = self.bounds[i][0] + binary_value
+
+    def crossover(self):
+        """perform crossover between individual parents"""
+        self.offspring_population[:, :] = self.parent_population[:, :]
+
+        # mate conscutive pairs of parents (0, 1), (2, 3), ...
+        # The population is shuffled so this does not need to be randomized
+        for i in range(int(self.population_size / 2)):
+            # trade bits in the offspring
+            crossover_arr = np.random.rand(self.nbits)
+            for j in range(self.nbits):
+                if crossover_arr[j] < self.crossover_rate:
+                    self.offspring_population[2 * i][j], \
+                        self.offspring_population[2 * i + 1][j] = \
+                        self.offspring_population[2 * i + 1][j], \
+                        self.offspring_population[2 * i][j]
+
+    def mutate(self):
+        """randomly mutate bits of each chromosome"""
+        for i in range(int(self.population_size)):
+            # mutate bits in the offspring
+            mutate_arr = np.random.rand(self.nbits)
+            for j in range(self.nbits):
+                if mutate_arr[j] < self.mutation_rate:
+                    self.offspring_population[i][j] = \
+                        (self.offspring_population[i][j] + 1) % 2
+
+    def optimize_ga(self):
+        """run the genetic algorithm"""
+
+        self.initialize_design_variables()
+        self.initialize_bits()
+        if self.population_size % 2 == 1:
+            self.population_size += 1
+        self.initialize_population()
+        self.initialize_fitness()
+
+        converged = False
+        ngens = 1
+        generation = 1
+        difference = self.tol * 10000.0
+        self.solution_history = np.zeros(self.max_generation + 1)
+        self.solution_history[0] = np.min(self.parent_fitness)
+
+        run_time = 0.0
+        start_time = time.time()
+        while converged is False and ngens < self.max_generation and \
```

Comment:
Do you want to add any log statements to this module? See other reV modules for examples of how to set up a logger object for a module. Only add debug level logs because this module is called in parallel and info level would get crazy.

Reply:
Yeah this is a good idea. I'll address this early next week